app/lib/study_sync_service.rb

```diff
@@ -1,5 +1,243 @@
+      # skip validation as we don't wont to set the acl in Terra as it already exists
+      new_share.save(validate: false)
+      new_share
+    elsif portal_permissions[acl_email] != StudyShare::PORTAL_ACL_MAP[acl_permissions['accessLevel']] &&
+          acl_email != study.user.email
+      # share exists, but permissions are wrong
+      share = study.study_shares.detect { |s| s.email == acl_email }
+      share.update(permission: StudyShare::PORTAL_ACL_MAP[acl_permissions['accessLevel']])
+      share
+    else
+      # permissions are correct, skip
+      nil
+    end
+  end
+
+  # main method to iterate list of remote files and create entries as needed
+  #
+  # * *params*
+  #   - +study+ (Study) => study to process remote files for
+  #
+  # * *returns*
+  #   - (Array<StudyFile>) => array of unsynced StudyFile documents
+  def self.process_all_remotes(study)
+    Rails.logger.info "processing all remotes for #{study.accession}"
+    workspace_files = ApplicationController.firecloud_client.execute_gcloud_method(
+      :get_workspace_files, 0, study.bucket_id, delimiter: '_scp_internal'
+    )
+    # we need to duplicate the workspace_files list so that we don't lose track of next?
+    file_extension_map = create_file_map(workspace_files.dup)
+    unsynced_files = process_file_batch(study, workspace_files, file_extension_map: file_extension_map)
+    while workspace_files.next?
+      Rails.logger.info "processing next batch of remotes for #{study.accession}"
+      workspace_files = workspace_files.next
+      unsynced_files += process_file_batch(study, workspace_files, file_extension_map: file_extension_map)
+    end
+    unsynced_files
+  end
+
+  # process a block of files from bucket
+  # will ignore submission outputs, files in _scp_internal or parse_logs
+  # will also add files to DirectoryListing objects as needed
+  #
+  # * *params*
+  #   - +study+ (Study) => study to create new entities in
+  #   - +files+ (Google::Cloud::Storage::File::List) => current batch of remote files in GCP bucket (up to 1K)
+  #   - +file_extension_map+ (Hash) => output from StudySyncService,create_file_map
+  #
+  # * *returns*
+  #   - (Array<StudyFile>) => array of unsynced StudyFile documents from batch
+  def self.process_file_batch(study, files, file_extension_map:)
+    unsynced_study_files = []
+    valid_files = remove_submission_outputs(study, files)
+    new_files = remove_synced_files(study, valid_files)
+    dir_files = find_files_for_directories(new_files, file_extension_map)
+    add_files_to_directories(study, dir_files)
+    unsynced_files_in_batch = remove_directory_files(dir_files, new_files)
+    Rails.logger.info "found #{unsynced_files_in_batch.count} remotes in batch to process for #{study.accession}"
+    unsynced_files_in_batch.each do |file|
+      next if file.size == 0
+
+      unsynced_file = StudyFile.new(study_id: study.id, name: file.name, upload_file_name: file.name,
```

Comment:
Out of curiosity, do you suspect there'd be any significant performance improvement by batching study file creation?  It's a reflexive thought that occurs to me whenever I see non-trivial tasks executing in each iteration.

Reply:
No, because this is calling `new` and not `create`, so the objects are not persisted yet.  Calling `StudyFile.new` returns essentially immediately as no callbacks have fired yet.  Good thought though!

Reply:
> this is calling new and not create, so the objects are not persisted yet

Thanks, today I learned!